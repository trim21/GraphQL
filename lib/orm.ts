import dayjs from 'dayjs';
import * as php from 'php-serialize';

import type { TopicDisplay } from './auth/rule';
import { UnexpectedNotFoundError } from './errors';
import { logger } from './logger';
import prisma from './prisma';
import type { ReplyState } from './topic';

export interface Page {
  limit?: number;
  offset?: number;
}

export interface IUser {
  id: number;
  username: string;
  nickname: string;
  groupID: number;
  img: string;
  regTime: number;
  sign: string;
}

export async function fetchUserByUsername(username: string): Promise<IUser | null> {
  const user = await prisma.members.findFirst({
    where: { username },
  });

  if (!user) {
    return null;
  }

  return {
    id: user.id,
    nickname: user.nickname,
    username: user.username,
    img: user.avatar,
    groupID: user.groupid,
    regTime: user.regdate,
    sign: user.sign,
  };
}

export async function fetchUser(userID: number): Promise<IUser | null> {
  if (!userID) {
    throw new Error(`undefined user id ${userID}`);
  }
  const user = await prisma.members.findFirst({
    where: { id: userID },
  });

  if (!user) {
    return null;
  }

  return {
    id: user.id,
    nickname: user.nickname,
    username: user.username,
    img: user.avatar,
    groupID: user.groupid,
    regTime: user.regdate,
    sign: user.sign,
  };
}

export interface Permission {
  app_erase?: boolean;
  ban_post?: boolean;
  ban_visit?: boolean;
  doujin_subject_erase?: boolean;
  doujin_subject_lock?: boolean;
  ep_edit?: boolean;
  ep_erase?: boolean;
  ep_lock?: boolean;
  ep_merge?: boolean;
  ep_move?: boolean;
  manage_app?: boolean;
  manage_report?: boolean;
  manage_topic_state?: boolean;
  manage_user?: boolean;
  manage_user_group?: boolean;
  manage_user_photo?: boolean;
  mono_edit?: boolean;
  mono_erase?: boolean;
  mono_lock?: boolean;
  mono_merge?: boolean;
  report?: boolean;
  subject_cover_erase?: boolean;
  subject_cover_lock?: boolean;
  subject_edit?: boolean;
  subject_erase?: boolean;
  subject_lock?: boolean;
  subject_merge?: boolean;
  subject_refresh?: boolean;
  subject_related?: boolean;
  user_ban?: boolean;
  user_group?: boolean;
  user_list?: boolean;
  user_wiki_apply?: boolean;
  user_wiki_approve?: boolean;
}

export async function fetchPermission(userGroup: number): Promise<Readonly<Permission>> {
  const permission = await prisma.userGroups.findFirst({ where: { usr_grp_id: userGroup } });
  if (!permission) {
    logger.warn("can't find permission for userGroup %d", userGroup);
    return {};
  }

  if (!permission.usr_grp_perm) {
    return {};
  }

  return Object.freeze(
    Object.fromEntries(
      Object.entries(
        php.unserialize(permission.usr_grp_perm) as Record<keyof Permission, string>,
      ).map(([key, value]) => [key, value === '1']),
    ),
  );
}

export async function addCreator<T extends { creatorID: number }>(
  arr: T[],
): Promise<(T & { creator: IUser })[]> {
  const users = await fetchUsers(arr.map((x) => x.creatorID));

  return arr.map((o) => {
    const user = users[o.creatorID];
    if (!user) {
      throw new UnexpectedNotFoundError(`user ${o.creatorID}`);
    }

    return { ...o, creator: user };
  });
}

export async function fetchUsers(userIDs: number[]): Promise<Record<number, IUser>> {
  if (userIDs.length === 0) {
    return {};
  }

  const users = await prisma.members.findMany({
    where: { id: { in: userIDs } },
  });

  return Object.fromEntries(
    users.map((user) => [
      user.id,
      {
        id: user.id,
        nickname: user.nickname,
        username: user.username,
        img: user.avatar,
        groupID: user.groupid,
        regTime: user.regdate,
        sign: user.sign,
        user_group: user.groupid,
      },
    ]),
  );
}

export interface ITopic {
  id: number;
  parentID: number;
  creatorID: number;
  updatedAt: number;
  createdAt: number;
  title: string;
  repliesCount: number;
}

export async function fetchTopicList(
  type: 'group' | 'subject',
  id: number,
  { limit = 30, offset = 0 }: Page,
  { display }: { display: TopicDisplay[] },
): Promise<[number, ITopic[]]> {
  if (type === 'group') {
    const where = {
      gid: id,
      display: { in: display },
    } as const;

    const total = await prisma.groupTopics.count({ where });
    const topics = await prisma.groupTopics.findMany({
      where,
      orderBy: { dateline: 'desc' },
      skip: offset,
      take: limit,
    });

    return [
      total,
      topics.map((x) => {
        return {
          id: x.id,
          parentID: x.gid,
          creatorID: x.uid,
          title: x.title,
          createdAt: x.dateline,
          updatedAt: x.lastpost,
          repliesCount: x.replies,
        };
      }),
    ];
  }

  const where = { sbj_tpc_subject_id: id, sbj_tpc_display: { in: display } } as const;
  const total = await prisma.chii_subject_topics.count({ where });

  const topics = await prisma.chii_subject_topics.findMany({
    where,
    orderBy: { sbj_tpc_dateline: 'desc' },
    skip: offset,
    take: limit,
  });

  return [
    total,
    topics.map((x) => {
      return {
        id: x.sbj_tpc_id,
        parentID: x.sbj_tpc_subject_id,
        creatorID: x.sbj_tpc_uid,
        title: x.sbj_tpc_title,
        updatedAt: x.sbj_tpc_dateline,
        createdAt: x.sbj_tpc_lastpost,
        repliesCount: x.sbj_tpc_replies,
      };
    }),
  ];
}

export async function fetchSubject(id: number) {
  const subject = await prisma.subjects.findFirst({
    where: { subject_id: id },
  });

  if (!subject) {
    return null;
  }

  const f = await prisma.subjectFields.findFirst({
    where: { subject_id: id },
  });

  if (!f) {
    throw new UnexpectedNotFoundError(`subject fields ${id}`);
  }

  return {
    id: subject.subject_id,
    nsfw: subject.subject_nsfw,
    redirect: f.field_redirect,
  };
}

interface IGroup {
  id: number;
  name: string;
  nsfw: boolean;
  description: string;
  title: string;
  createdAt: number;
  totalMembers: number;
  icon: string;
  accessible: boolean;
}

export async function fetchGroupByID(id: number): Promise<IGroup | null> {
  const group = await prisma.chii_groups.findFirst({
    where: { grp_id: id },
  });

  if (!group) {
    return null;
  }

  return {
    id: group.grp_id,
    name: group.grp_name,
    title: group.grp_title,
    nsfw: group.grp_nsfw,
    description: group.grp_desc,
    createdAt: group.grp_builddate,
    icon: group.grp_icon,
    totalMembers: group.grp_members,
    accessible: group.grp_accessible,
  } satisfies IGroup;
}

export async function fetchGroup(name: string): Promise<IGroup | null> {
  const group = await prisma.chii_groups.findFirst({
    where: { grp_name: name },
  });

  if (!group) {
    return null;
  }

  return {
    id: group.grp_id,
    name: group.grp_name,
    title: group.grp_title,
    nsfw: group.grp_nsfw,
    description: group.grp_desc,
    icon: group.grp_icon,
    createdAt: group.grp_builddate,
    totalMembers: group.grp_members,
    accessible: group.grp_accessible,
  } satisfies IGroup;
}

interface IBaseReply {
  id: number;
  text: string;
  creatorID: number;
  state: number;
  createdAt: number;
}

interface ISubReply extends IBaseReply {
  repliedTo: number;
}

export interface IReply extends IBaseReply {
  replies: ISubReply[];
}

interface ITopicDetails {
  id: number;
  title: string;
  text: string;
  state: number;
  createdAt: number;
  creatorID: number;
  // group ID or subject ID
  parentID: number;
  replies: IReply[];
}

export async function fetchTopicDetails(type: 'group', id: number): Promise<ITopicDetails | null> {
  const topic = await prisma.groupTopics.findFirst({
    where: { id: id },
  });

  if (!topic) {
    return null;
  }

  const replies = await prisma.groupPosts.findMany({
    where: {
      mid: topic.id,
    },
  });

  const top = replies.shift();
  if (!top) {
    throw new UnexpectedNotFoundError(`top reply of topic(${type}) ${id}`);
  }

  const subReplies: Record<number, ISubReply[]> = {};

  for (const x of replies.filter((x) => x.related !== 0)) {
    const sub: ISubReply = {
      id: x.id,
      repliedTo: x.related,
      creatorID: x.uid,
      text: x.content,
      state: x.state,
      createdAt: x.dateline,
    };

    subReplies[x.related] ??= [];
    subReplies[x.related]?.push(sub);
  }

  const topLevelReplies = replies
    .filter((x) => x.related === 0)
    .map(function (x): IReply {
      return {
        id: x.id,
        replies: subReplies[x.id] ?? ([] as ISubReply[]),
        creatorID: x.uid,
        text: x.content,
        state: x.state,
        createdAt: x.dateline,
      };
    });

  return {
    id: topic.id,
    title: topic.title,
    parentID: topic.gid,
    text: top.content,
    state: topic.state,
    replies: topLevelReplies,
    creatorID: top.uid,
    createdAt: top.dateline,
  } satisfies ITopicDetails;
}

export async function fetchFriends(id?: number): Promise<Record<number, boolean>> {
  if (!id) {
    return {};
  }

  const friends = await prisma.friends.findMany({
    where: { frd_uid: id },
  });

  return Object.fromEntries(friends.map((x) => [x.frd_fid, true]));
}

interface PostCreation {
  title: string;
  content: string;
  groupID: number;
  userID: number;
  display: TopicDisplay;
  state: ReplyState;
}

export async function createPostInGroup(post: PostCreation): Promise<{ id: number }> {
  const now = dayjs();

  return await prisma.$transaction(async (t) => {
    const topic = await t.groupTopics.create({
      data: {
        title: post.title,
        gid: post.groupID,
        uid: post.userID,
        state: post.state,
        lastpost: now.unix(),
        dateline: now.unix(),
        replies: 0,
        display: post.display,
      },
    });

    await t.groupPosts.create({
      data: {
        mid: topic.id,
        dateline: now.unix(),
        state: post.state,
        uid: post.userID,
        content: post.content,
        related: 0,
      },
    });

    return { id: topic.id };
  });
}

export async function isMemberInGroup(gid: number, uid: number): Promise<boolean> {
  const inGroup = await prisma.groupMembers.findFirst({
    where: { gmb_gid: gid, gmb_uid: uid },
  });

  return Boolean(inGroup);
}

export async function fetchUserX(id: number): Promise<IUser> {
  const u = await fetchUser(id);
  if (!u) {
    throw new UnexpectedNotFoundError(`user ${id}`);
  }

  return u;
}
