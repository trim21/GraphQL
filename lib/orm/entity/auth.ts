import { Entity, PrimaryKey, Property } from '@mikro-orm/core';

@Entity({ tableName: 'chii_oauth_access_tokens', schema: 'bangumi' })
export class OauthAccessTokens {
  @PrimaryKey({ type: 'mediumint', name: 'id', autoincrement: true })
  id!: number;

  @Property({ name: 'type', unsigned: true, default: "'0'", type: 'tinyint' })
  type!: number;

  @Property({ name: 'access_token', unique: true, length: 40, type: 'varchar' })
  accessToken!: string;

  @Property({ name: 'client_id', length: 80, type: 'varchar' })
  clientId!: string;

  @Property({ name: 'user_id', nullable: true, length: 80, type: 'varchar' })
  userId!: string;

  @Property({ name: 'expires', default: 'CURRENT_TIMESTAMP', type: 'timestamp' })
  expires!: Date;

  @Property({ name: 'scope', nullable: true, length: 4000, type: 'varchar' })
  scope: string | undefined;

  @Property({ name: 'info', length: 255, type: 'varchar' })
  info!: string;
}

@Entity({ schema: 'bangumi', tableName: 'chii_os_web_sessions' })
export class WebSessions {
  @Property({
    primary: true,
    name: 'key',
    type: 'char',
    comment: 'session key',
    length: 64,
  })
  key!: string;

  @Property({ name: 'user_id', comment: 'uint32 user id', unsigned: true, type: 'int' })
  userID!: number;

  @Property({ name: 'value', comment: 'json encoded session data', type: 'mediumblob' })
  value!: Buffer;

  @Property({
    name: 'created_at',
    type: 'bigint',
    comment: 'int64 unix timestamp, when session is created',
  })
  createdAt!: number;

  @Property({
    name: 'expired_at',
    type: 'bigint',
    comment: 'int64 unix timestamp, when session is expired',
  })
  expiredAt!: number;
}
