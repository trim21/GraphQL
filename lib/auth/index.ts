import { createError } from '@fastify/error';
import dayjs from 'dayjs';
import NodeCache from 'node-cache';

import { redisPrefix } from '../config';
import type { IUser, Permission } from '../orm';
import { fetchPermission, fetchUser } from '../orm';
import prisma from '../prisma';
import redis from '../redis';

const tokenPrefix = 'Bearer ';
export const NeedLoginError = createError('NEED_LOGIN', 'you need to login before %s', 401);
export const NotAllowedError = createError('NEED_LOGIN', `you don't have permission to %s`, 401);
const HeaderInvalidError = createError('AUTHORIZATION_INVALID', '%s', 401);
const TokenNotValidError = createError(
  'TOKEN_INVALID',
  "can't find access token or it has been expired",
  401,
);
const MissingUserError = createError('MISSING_USER', "can't find user", 500);

export const enum UserGroup {
  Unknown = 0,
  Admin = 1,
  BangumiAdmin,
  WindowAdmin,
  Quite,
  Banned,
  // 不太清除具体是什么
  _6,
  // 不太清除具体是什么
  _7,
  CharacterAdmin,
  WikiAdmin,
  Normal,
  WikiEditor,
}

export interface IAuth {
  userID: number;
  login: boolean;
  allowNsfw: boolean;
  permission: Readonly<Permission>;
  groupID: UserGroup;
}

export async function byHeader(key: string | string[] | undefined): Promise<IAuth | null> {
  if (!key) {
    return emptyAuth();
  }

  if (Array.isArray(key)) {
    throw new HeaderInvalidError("can't providing multiple access token");
  }

  if (!key.startsWith(tokenPrefix)) {
    throw new HeaderInvalidError('authorization header should have "Bearer ${TOKEN}" format');
  }

  const token = key.slice(tokenPrefix.length);
  if (!token) {
    throw new HeaderInvalidError('authorization header missing token');
  }

  return await byToken(token);
}

export async function byToken(access_token: string): Promise<IAuth | null> {
  const key = `${redisPrefix}-auth-access-token-${access_token}`;
  const cached = await redis.get(key);
  if (cached) {
    const user = JSON.parse(cached) as IUser;
    return await userToAuth(user);
  }

  const token = await prisma.chii_oauth_access_tokens.findFirst({
    where: { access_token, expires: { gte: new Date() } },
  });

  if (!token) {
    throw new TokenNotValidError();
  }

  const u = await fetchUser(Number.parseInt(token.user_id));
  if (!u) {
    throw new MissingUserError();
  }

  await redis.set(key, JSON.stringify(u), 'EX', 60 * 60 * 24);

  return await userToAuth(u);
}

export async function byUserID(userID: number): Promise<IAuth> {
  const key = `${redisPrefix}-auth-user-id-${userID}`;
  const cached = await redis.get(key);
  if (cached) {
    return await userToAuth(JSON.parse(cached) as IUser);
  }

  const u = await fetchUser(userID);
  if (!u) {
    throw new MissingUserError();
  }

  await redis.set(key, JSON.stringify(u), 'EX', 60 * 60 * 24);

  return await userToAuth(u);
}

const permissionCache = new NodeCache({ stdTTL: 60 * 10 });

async function getPermission(userGroup?: number): Promise<Readonly<Permission>> {
  if (!userGroup) {
    return {};
  }

  const cached = permissionCache.get(userGroup);
  if (cached) {
    return cached;
  }

  const p = await fetchPermission(userGroup);

  permissionCache.set(userGroup, p);

  return p;
}

export function emptyAuth(): IAuth {
  return {
    userID: 0,
    login: false,
    permission: {},
    allowNsfw: false,
    groupID: 0,
  };
}

async function userToAuth(user: IUser): Promise<IAuth> {
  return {
    userID: user.id,
    login: true,
    permission: await getPermission(user.groupID),
    allowNsfw: user.regTime - dayjs().unix() <= 60 * 60 * 24 * 90,
    groupID: user.groupID,
  };
}
