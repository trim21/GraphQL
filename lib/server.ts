import AltairFastify from 'altair-fastify-plugin';
import type { FastifyInstance, FastifyRequest, FastifyServerOptions } from 'fastify';
import { fastify } from 'fastify';
import metricsPlugin from 'fastify-metrics';
import mercurius from 'mercurius';
import { register } from 'prom-client';

import { emptyAuth } from './auth';
import * as auth from './auth';
import { production, stage, testing } from './config';
import type { Context } from './graphql/context';
import { schema } from './graphql/schema';
import prisma from './prisma';
import * as rest from './rest';

declare module 'fastify' {
  interface FastifyRequest {
    // use clientIp privided by cloudflare
    clientIP: string;
  }
}

export async function createServer(opts: FastifyServerOptions = {}): Promise<FastifyInstance> {
  if (production || stage) {
    opts.requestIdHeader ??= 'cf-ray';
  }

  const server = fastify(opts);

  server.decorateRequest('clientIP', '');

  // eslint-disable-next-line @typescript-eslint/require-await
  server.addHook('preHandler', async (req) => {
    const cfClientIp = req.headers['cf-connecting-ip'] as string | undefined;
    if (cfClientIp) {
      req.clientIP = cfClientIp;
      return;
    }

    req.clientIP = req.ip;
  });

  if (!testing) {
    server.get('/metrics', async (_req, res) => {
      const prismaMetrics = await prisma.$metrics.prometheus();
      const appMetrics = await register.metrics();
      return res.send(appMetrics + prismaMetrics);
    });

    await server.register(metricsPlugin, {
      endpoint: null,
      routeMetrics: {
        groupStatusCodes: true,
        overrides: {
          histogram: {
            buckets: [0.05, 0.1, 0.3, 0.5, 0.75, 1, 2, 3],
          },
        },
      },
    });
  }

  await server.register(mercurius, {
    schema,
    path: '/v0/graphql',
    graphiql: false,
    allowBatchedQueries: true,
    context: async (request: FastifyRequest): Promise<Context> => {
      const a = await auth.byHeader(request.headers.authorization);
      if (a) {
        return { prisma, auth: a };
      }

      return { prisma, auth: emptyAuth() };
    },
  });

  await server.register(AltairFastify, {
    path: '/v0/altair/',
    baseURL: '/v0/altair/',
    endpointURL: '/v0/graphql',
    initialSettings: {
      theme: 'dark',
      'plugin.list': ['altair-graphql-plugin-graphql-explorer'],
    },
  });

  await server.register(rest.setup);

  return server;
}
