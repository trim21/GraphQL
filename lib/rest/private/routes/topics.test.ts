import dayjs from 'dayjs';
import { fastify } from 'fastify';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';

import type { IAuth } from '../../../auth';
import { UserGroup } from '../../../auth';
import * as orm from '../../../orm';
import { createServer } from '../../../server';
import { ReplyState } from '../../../topic';
import * as Topic from '../../../topic';
import * as topicAPI from './topics';

function createServerWithAuth(auth: IAuth) {
  const app = fastify();

  app.addHook('preHandler', (req, res, done) => {
    req.auth = auth;
    done();
  });

  return app;
}

describe('group topics', () => {
  test('should failed on not found group', async () => {
    const app = await createServer();

    const res = await app.inject({
      url: '/p1/groups/non-existing/topics',
    });

    expect(res.statusCode).toBe(404);
    expect(res.json()).toMatchSnapshot();
  });

  test('should return data', async () => {
    const app = await createServer();

    const res = await app.inject({
      url: '/p1/groups/sandbox/topics',
    });

    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should fetch group profile', async () => {
    const app = await createServer();

    const res = await app.inject('/p1/groups/sandbox/profile');
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should fetch topic details', async () => {
    const app = await createServer();

    const res = await app.inject('/p1/groups/-/topics/371602');
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });
});

describe('create group post', () => {
  const createPostInGroup = vi.fn().mockResolvedValue({ id: 1 });
  vi.spyOn(orm, 'createPostInGroup').mockImplementation(createPostInGroup);

  test('should create group topic', async () => {
    const app = fastify();

    app.addHook('preHandler', (req, res, done) => {
      req.auth = {
        groupID: UserGroup.Normal,
        login: true,
        permission: {},
        allowNsfw: true,
        userID: 100,
      } satisfies IAuth;
      done();
    });

    await app.register(topicAPI.setup);

    const res = await app.inject({
      url: '/groups/sandbox/topics',
      method: 'post',
      payload: {
        title: 'post title',
        content: 'post contents',
      },
    });

    expect(res.json()).toMatchObject({ id: 1 });
    expect(res.statusCode).toBe(200);
  });

  test('should not create with banned user', async () => {
    const app = fastify();

    app.addHook('preHandler', (req, res, done) => {
      req.auth = {
        groupID: UserGroup.Normal,
        login: true,
        permission: {
          ban_post: true,
        },
        allowNsfw: true,
        userID: 1,
      } satisfies IAuth;
      done();
    });

    await app.register(topicAPI.setup);

    const res = await app.inject({
      url: '/groups/sandbox/topics',
      method: 'post',
      payload: {
        title: 'post title',
        content: 'post contents',
      },
    });

    expect(res.statusCode).toBe(401);
    expect(createPostInGroup).toBeCalledTimes(1);
  });
});

describe('create group post reply', () => {
  const createTopicReply = vi.fn().mockResolvedValue({
    id: 6,
    content: '',
    state: ReplyState.Normal,
    createdAt: dayjs('2021-10-21').unix(),
    type: Topic.Type.group,
    topicID: 371602,
    user: {
      img: '',
      username: 'u',
      groupID: UserGroup.Normal,
      id: 9,
      nickname: 'n',
      regTime: dayjs('2008-10-01').unix(),
      sign: '',
    },
  });
  const getPostMock = vi.fn().mockResolvedValue({
    id: 6,
    content: '',
    state: ReplyState.Normal,
    createdAt: dayjs('2021-10-21').unix(),
    type: Topic.Type.group,
    topicID: 371602,
    user: {
      img: '',
      username: 'u',
      groupID: UserGroup.Normal,
      id: 9,
      nickname: 'n',
      regTime: dayjs('2008-10-01').unix(),
      sign: '',
    },
  });

  beforeEach(() => {
    vi.spyOn(Topic, 'getPost').mockImplementation(getPostMock);
    vi.spyOn(Topic, 'createTopicReply').mockImplementation(createTopicReply);
    vi.spyOn(orm, 'fetchTopicDetails').mockImplementationOnce(
      (type: 'group', id: number): ReturnType<typeof orm['fetchTopicDetails']> => {
        if (id !== 371602) {
          return Promise.resolve(null);
        }

        return Promise.resolve({
          replies: [],
          creatorID: 0,
          id: id,
          title: 't',
          createdAt: dayjs().unix(),
          text: 't',
          state: ReplyState.Normal,
          parentID: 1,
        });
      },
    );
  });

  afterEach(() => {
    vi.resetModules();
  });

  test('should create group post reply', async () => {
    const app = createServerWithAuth({
      groupID: UserGroup.Normal,
      login: true,
      permission: {},
      allowNsfw: true,
      userID: 100,
    });

    await app.register(topicAPI.setup);

    const res = await app.inject({
      url: '/groups/-/topics/371602/replies',
      method: 'post',
      payload: {
        content: 'post contents',
      },
    });

    expect(res.statusCode).toBe(200);
  });

  test('should not create with banned user', async () => {
    const app = createServerWithAuth({
      groupID: UserGroup.Normal,
      login: true,
      permission: {
        ban_post: true,
      },
      allowNsfw: true,
      userID: 1,
    });

    await app.register(topicAPI.setup);

    const res = await app.inject({
      url: '/groups/-/topics/371602/replies',
      method: 'post',
      payload: {
        content: 'post contents',
      },
    });

    expect(res.statusCode).toBe(401);
    expect(createTopicReply).toBeCalledTimes(1);
  });

  test('should not create on non-existing topic', async () => {
    const app = createServerWithAuth({
      groupID: UserGroup.Normal,
      login: true,
      permission: {},
      allowNsfw: true,
      userID: 1,
    });

    await app.register(topicAPI.setup);

    const res = await app.inject({
      url: '/groups/-/topics/3716000/replies',
      method: 'post',
      payload: {
        content: 'post contents',
      },
    });

    expect(res.statusCode).toBe(404);
    expect(res.json()).toMatchSnapshot();
  });

  test('should not create on non-existing topic reply', async () => {
    const app = createServerWithAuth({
      groupID: UserGroup.Normal,
      login: true,
      permission: {},
      allowNsfw: true,
      userID: 1,
    });

    await app.register(topicAPI.setup);

    const res = await app.inject({
      url: '/groups/-/topics/371602/replies',
      method: 'post',
      payload: {
        content: 'post contents',
        relatedID: 11,
      },
    });

    expect(res.statusCode).toBe(404);
    expect(res.json()).toMatchSnapshot();
  });
});
