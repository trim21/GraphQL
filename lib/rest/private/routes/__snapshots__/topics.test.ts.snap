// Vitest Snapshot v1

exports[`create group post reply > should not create on non-existing topic 1`] = `
{
  "code": "NOT_FOUND",
  "error": "Not Found",
  "message": "topic 3716000 not found",
  "statusCode": 404,
}
`;

exports[`create group post reply > should not create on non-existing topic reply 1`] = `
{
  "code": "NOT_FOUND",
  "error": "Not Found",
  "message": "parent topic id 11 not found",
  "statusCode": 404,
}
`;

exports[`group topics > should failed on not found group 1`] = `
{
  "code": "NOT_FOUND",
  "error": "Not Found",
  "message": "group not found",
  "statusCode": 404,
}
`;

exports[`group topics > should fetch group profile 1`] = `
{
  "group": {
    "createdAt": 1531631310,
    "description": "[s]非[/s]官方沙盒",
    "icon": "https://lain.bgm.tv/pic/icon/s/000/00/42/4215.jpg?r=1531631345",
    "id": 4215,
    "name": "sandbox",
    "nsfw": false,
    "title": "沙盒",
    "totalMembers": 3,
  },
  "inGroup": false,
  "recentAddedMembers": [
    {
      "avatar": {
        "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
        "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
        "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
      },
      "id": 382951,
      "joinedAt": 1640499538,
      "nickname": "树洞酱",
      "username": "382951",
    },
    {
      "avatar": {
        "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
        "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
        "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
      },
      "id": 427613,
      "joinedAt": 1531631570,
      "nickname": "nickname 427613",
      "username": "427613",
    },
    {
      "avatar": {
        "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
        "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
        "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
      },
      "id": 287622,
      "joinedAt": 1531631310,
      "nickname": "nickname 287622",
      "username": "287622",
    },
  ],
  "topics": [
    {
      "createdAt": 1657885648,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 287622,
        "nickname": "nickname 287622",
        "sign": "sing 287622",
        "user_group": 0,
        "username": "287622",
      },
      "id": 371602,
      "parentID": 4215,
      "repliesCount": 2,
      "title": "tes",
      "updatedAt": 1662283112,
    },
  ],
  "totalTopics": 1,
}
`;

exports[`group topics > should fetch topic details 1`] = `
{
  "createdAt": 1657885648,
  "creator": {
    "avatar": {
      "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
      "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
      "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
    },
    "id": 287622,
    "nickname": "nickname 287622",
    "sign": "sing 287622",
    "user_group": 0,
    "username": "287622",
  },
  "group": {
    "createdAt": 1531631310,
    "description": "[s]非[/s]官方沙盒",
    "icon": "https://lain.bgm.tv/pic/icon/s/000/00/42/4215.jpg?r=1531631345",
    "id": 4215,
    "name": "sandbox",
    "nsfw": false,
    "title": "沙盒",
    "totalMembers": 3,
  },
  "id": 371602,
  "replies": [
    {
      "createdAt": 1658030918,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 287622,
        "nickname": "nickname 287622",
        "sign": "sing 287622",
        "user_group": 0,
        "username": "287622",
      },
      "id": 2056216,
      "isFriend": false,
      "replies": [],
      "state": 6,
      "text": "",
    },
    {
      "createdAt": 1662276469,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 287622,
        "nickname": "nickname 287622",
        "sign": "sing 287622",
        "user_group": 0,
        "username": "287622",
      },
      "id": 2092004,
      "isFriend": false,
      "replies": [],
      "state": 7,
      "text": "",
    },
    {
      "createdAt": 1662282514,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 1,
        "nickname": "nickname 1",
        "sign": "sing 1",
        "user_group": 0,
        "username": "1",
      },
      "id": 2092065,
      "isFriend": false,
      "replies": [],
      "state": 5,
      "text": "",
    },
    {
      "createdAt": 1662282518,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 1,
        "nickname": "nickname 1",
        "sign": "sing 1",
        "user_group": 0,
        "username": "1",
      },
      "id": 2092066,
      "isFriend": false,
      "replies": [],
      "state": 1,
      "text": "",
    },
    {
      "createdAt": 1662282523,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 1,
        "nickname": "nickname 1",
        "sign": "sing 1",
        "user_group": 0,
        "username": "1",
      },
      "id": 2092067,
      "isFriend": false,
      "replies": [],
      "state": 2,
      "text": "",
    },
    {
      "createdAt": 1662282554,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 287622,
        "nickname": "nickname 287622",
        "sign": "sing 287622",
        "user_group": 0,
        "username": "287622",
      },
      "id": 2092069,
      "isFriend": false,
      "replies": [
        {
          "createdAt": 1662283112,
          "creator": {
            "avatar": {
              "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
              "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
              "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
            },
            "id": 287622,
            "nickname": "nickname 287622",
            "sign": "sing 287622",
            "user_group": 0,
            "username": "287622",
          },
          "id": 2092074,
          "isFriend": false,
          "state": 0,
          "text": "sub",
        },
      ],
      "state": 0,
      "text": "q",
    },
  ],
  "state": 2,
  "text": "tt",
  "title": "tes",
}
`;

exports[`group topics > should return data 1`] = `
{
  "data": [
    {
      "createdAt": 1657885648,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 287622,
        "nickname": "nickname 287622",
        "sign": "sing 287622",
        "user_group": 0,
        "username": "287622",
      },
      "id": 371602,
      "parentID": 4215,
      "repliesCount": 2,
      "title": "tes",
      "updatedAt": 1662283112,
    },
  ],
  "total": 1,
}
`;
